In [1]:
# Imports and pip installations (if needed)
%pip install pandas

import pandas as pd
pd.set_option('display.max_columns', None) # display all columns in terminal

Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical_dataset = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_dataset = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(categorical_dataset)
# print(categorical_dataset.iloc[0])
print()
print(numerical_dataset)
# print(numerical_dataset.iloc[0])


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
The dataset is about different features that can be used to predict chronic kidney disease.
The dataset has 24 features (11 numeric and 13 categorical), and 1 target label (class), which can be either "ckd" or "notckd".
Examples of categorical columns: Bacteria (ba), Hypertension (htn), and Diabetes Mellitus (dm).
Examples of numerical columns: Blood urea (bu), Potassium (pot), and White blood cell count (wc).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dup = numerical_dataset.duplicated().sum()
if num_dup > 0:
    print("The numerical dataset has duplicates")
    print("Number of duplicates: ", num_dup)
    num_dup_rows = numerical_dataset[numerical_dataset.duplicated("unique_id")]
    print("Duplicate rows in the numerical dataset:")
    print(num_dup_rows)
    # Drop these duplicate rows
    numerical_dataset.drop_duplicates("unique_id", inplace=True)
else:
    print("No duplicates in the numerical dataset")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_dup = categorical_dataset.duplicated().sum()
if cat_dup > 0:
    print("The categorical dataset has duplicates")
    print("Number of duplicates: ", cat_dup)
    cat_dup_rows = categorical_dataset[categorical_dataset.duplicated("unique_id")]
    print("Duplicate rows in the categorical dataset:")
    print(cat_dup_rows)
    categorical_dataset.drop_duplicates("unique_id", inplace=True)
else:
    print("No duplicates in the categorical dataset")

# Test
# print("Verify that duplicates have been removed")
# print(numerical_dataset.duplicated().sum())
# print(categorical_dataset.duplicated().sum())


The numerical dataset has duplicates
Number of duplicates:  30
Duplicate rows in the numerical dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: There are 390 unique ids that are present in both the numerical and categorical datasets. I think that this is because some rows with the same unique id have both numerical and categorical features. In my code, I have used an outer join to preserve the rows which appear in one dataset but not the other, and fill in the blank columns in the merged dataset with NaN.

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

dataset = pd.merge(numerical_dataset, categorical_dataset, on="unique_id", how="outer")
# print(len(dataset))
#Print the combined dataset
print(dataset)
# print(dataset.iloc[0])

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0   

        wbcc  rbcc   al   su     rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  no

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
num_missing_rows = dataset[dataset.isnull().any(axis=1)].shape[0]
print(num_missing_rows)
percent_missing = (num_missing_rows / len(dataset)) * 100

# Print %
print("Percentage of rows with atleast one missing value: ", percent_missing)

# Drop these rows from the dataset
dataset.dropna(inplace=True)

# Print the Dataset
print(dataset)

247
Percentage of rows with atleast one missing value:  61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su     rbc 

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
dataset.sort_values(by="Target", inplace=True)
dataset.reset_index(drop=True, inplace=True)

# Print the dataset
print(dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su       rbc        pc         pcc          ba  htn  \
0     3800.0   2.1  3.0  0.0  

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dataset = pd.get_dummies(data=dataset, columns=categorical_columns, dtype=int)

# Print the dataset
print(dataset)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
0     3800.0   2.1  3.0  0.0            

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [ ]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    mean = dataset[col].mean()
    stdev3 = dataset[col].std() * 3
    dataset = dataset[(dataset[col] >= (mean - stdev3)) & (dataset[col] <= (mean + stdev3))]

# Print the dataset
print(dataset)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0   

        wbcc  rbcc   al   su  rbc_abnormal  rbc_normal  pc_abnormal  \
2     9200.0   3.2  2.0  0.0            

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [ ]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    dataset[col] = (dataset[col] - dataset[col].mean()) / dataset[col].std()

# Print the dataset
print(dataset)

     unique_id       age        bp       bgr        bu        sc       sod  \
2       343710  0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087   
5       397388  0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053   
7       438182  1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910   
8       474407 -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013   
9       475200  0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124   
..         ...       ...       ...       ...       ...       ...       ...   
148     486397 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087   
149     203466 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910   
150     482105 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910   
151     228849  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124   
152     996076  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801   

          pot      hemo       pcv      wbcc      rbcc   al   su

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column is not appropriate for modeling and predictions, because it has no relation with whether the patient has CKD or not. The unique_id is just there to identify the patient's records, and it has no correlation to CKD. If the column is not removed, it would reduce the model's accuracy because the model would include it as a feature while predicting CKD despite it having no predictive value. Thus, the column for unique_id needs to be removed.

In [ ]:
#Remove that column
dataset = dataset.drop(columns=["unique_id"])

# Print the dataset
print(dataset)
# print(dataset.iloc[0])

          age        bp       bgr        bu        sc       sod       pot  \
2    0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087  1.367853   
5    0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053 -1.298281   
7    1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910  2.308841   
8   -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013 -1.298281   
9    0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124  0.426864   
..        ...       ...       ...       ...       ...       ...       ...   
148 -1.769750 -0.291121  0.036896  0.111307 -0.264117 -0.956087 -0.827787   
149 -1.644048 -1.310043 -0.537820 -0.860662 -0.129786 -0.429910 -0.043630   
150 -0.261328 -1.310043 -0.582029  0.186074 -0.599945 -0.429910 -0.200461   
151  0.618586  0.727802 -0.471507  0.335608 -0.129786 -0.079124 -1.298281   
152  0.555735 -1.310043 -0.360984  0.298224 -0.129786  1.674801  0.583696   

         hemo       pcv      wbcc      rbcc   al   su  rbc_abnormal  \
2   

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [11]:
# Export the dataframe to a new csv file
dataset.to_csv("preprocessed_dataset_csv", index=False)

# Export the dataframe to a new json file
dataset.to_json("preprocessed_dataset_json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [12]:
# Part 4.1 Code
print(dataset.head(15))

         age        bp       bgr        bu        sc       sod       pot  \
2   0.807138  0.727802  1.142120  3.999185  1.683685 -0.956087  1.367853   
5   0.932840  2.765646  3.507298 -0.225144  1.213526  0.447053 -1.298281   
7   1.561350  0.727802  2.910477  3.774884  2.153844 -0.429910  2.308841   
8  -1.706899  1.746724 -0.316775 -0.038227  0.206042 -2.710013 -1.298281   
9   0.995691  2.765646  0.921075  0.485141  3.900150 -0.079124  0.426864   
10  0.744287 -1.310043  3.684134 -0.187760  0.206042 -1.833050 -2.082438   
12  0.492884  1.746724  0.169523  2.466464  3.564322 -1.657657  0.740527   
13  1.309946 -0.291121  2.048403  2.055246  3.295659 -3.586975 -0.670955   
15 -0.135626 -1.310043  0.921075  1.905712  1.280691  0.096268 -0.514124   
18  0.869989 -0.291121  0.014792  0.036540  0.206042 -0.780695  0.583696   
19  0.430033  0.727802  2.048403  1.195427  1.683685 -0.605302  0.897359   
20  0.995691 -1.310043  2.601015  0.634675  1.952347 -0.605302  1.681516   
22  1.435648

### Paste the markdown table here

| age | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------------|------------|------------|------------|------------|------------|------------|-------------|-------------|-------------|--------------|--------|--------|----------------|--------------|----------------|-------------|------------------|----------------|------------------|--------------|----------|-----------|---------|----------|-------------|--------------|----------------|----------------|---------|----------|----------|-----------|----------------|-------------------|
| 0.807138 | 0.727802 | 1.142120 | 3.999185 | 1.683685 | -0.956087 | 1.367853 | -2.902248 | -2.894560 | 0.517701 | -2.293292 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | False | True | False | True | False | True | True | False |
| 0.932840 | 2.765646 | 3.507298 | -0.225144 | 1.213526 | 0.447053 | -1.298281 | -0.604893 | -0.588449 | 0.782525 | -1.080055 | 2.0 | 2.0 | False | True | False | True | True | False | False | True | False | True | True | False | False | True | True | False | True | False | True | False | True | False |
| 1.561350 | 0.727802 | 2.910477 | 3.774884 | 2.153844 | -0.429910 | 2.308841 | -1.688551 | -1.597372 | -0.365045 | -0.958731 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | False | True | False | True | True | False | True | False | True | False | True | False |
| -1.706899 | 1.746724 | -0.316775 | -0.038227 | 0.206042 | -2.710013 | -1.298281 | -2.642170 | -3.038692 | 1.930095 | -1.444026 | 4.0 | 0.0 | False | True | True | False | False | True | False | False | True | False | False | True | True | False | True | False | True | False | False | True | True | False |
| 0.995691 | 2.765646 | 0.921075 | 0.485141 | 3.900150 | -0.079124 | 0.426864 | -2.815556 | -2.606296 | -0.232633 | -2.050645 | 4.0 | 2.0 | True | False | True | False | True | False | True | False | False | True | True | False | True | False | True | False | False | True | True | False | True | False |
| 0.744287 | -1.310043 | 3.684134 | -0.187760 | 0.206042 | -1.833050 | -2.082438 | -2.815556 | -2.750428 | 3.165939 | -2.535939 | 3.0 | 1.0 | False | True | True | False | False | True | True | False | False | True | True | False | True | False | True | False | True | False | True | False | True | False |
| 0.492884 | 1.746724 | 0.169523 | 2.466464 | 3.564322 | -1.657657 | 0.740527 | -2.295400 | -2.173900 | -0.718144 | -2.050645 | 2.0 | 0.0 | True | False | True | False | True | False | True | False | False | True | True | False | True | False |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: To reduce LLM hallucinations during the process of large data conversion, we can use LLMs trained on domain specific knowledge. Two ways of doing this are fine-tuning and retrieval augmented generation (RAG). In addition, we can set guardrails for LLMs, which are some tests that detect hallucinations. I believe that the best way to avoid these situations are to use LLMs as a tool for helping us and not an agent that provides the correct output. I think that we should use LLMs to suggest short snippets of code and hints, and we should use this to design our own code to perform these data transformations. While doing section 4 of this project, I found that the respones of the LLM were not always correct, and I manually verified the LLM's output to make sure that the output did not contain any hallucinations.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
new_column_names = [
    'age',                           # age
    'blood_pressure',                # bp
    'blood_glucose_random',          # bgr
    'blood_urea',                    # bu
    'serum_creatinine',              # sc
    'sodium',                        # sod
    'potassium',                     # pot
    'hemoglobin',                    # hemo
    'packed_cell_volume',            # pcv
    'white_blood_cell_count',        # wbcc (added based on context)
    'red_blood_cell_count',          # rbcc (added based on context)
    'albumin',                       # al
    'sugar',                         # su
    'red_blood_cells_abnormal',      # rbc_abnormal
    'red_blood_cells_normal',        # rbc_normal
    'pus_cell_abnormal',             # pc_abnormal
    'pus_cell_normal',               # pc_normal
    'pus_cell_clumps_notpresent',    # pcc_notpresent
    'pus_cell_clumps_present',       # pcc_present
    'bacteria_not_present',          # ba_notpresent
    'bacteria_present',              # ba_present
    'hypertension_no',               # htn_no
    'hypertension_yes',              # htn_yes
    'diabetes_mellitus_no',          # dm_no
    'diabetes_mellitus_yes',         # dm_yes
    'coronary_artery_disease_no',    # cad_no
    'coronary_artery_disease_yes',   # cad_yes
    'appetite_good',                 # appet_good
    'appetite_poor',                 # appet_poor
    'pedal_edema_no',                # pe_no
    'pedal_edema_yes',               # pe_yes
    'anemia_no',                     # ane_no 
    'anemia_yes',                    # ane_yes
    'Target_chronic_kidney_disease', # Target_ckd
    'Target_not_chronic_kidney_disease' # Target_notckd
]
# print(len(current_columns))
# print(len(new_column_names))

dataset.columns = new_column_names
print(dataset)


          age  blood_pressure  blood_glucose_random  blood_urea  \
2    0.807138        0.727802              1.142120    3.999185   
5    0.932840        2.765646              3.507298   -0.225144   
7    1.561350        0.727802              2.910477    3.774884   
8   -1.706899        1.746724             -0.316775   -0.038227   
9    0.995691        2.765646              0.921075    0.485141   
..        ...             ...                   ...         ...   
148 -1.769750       -0.291121              0.036896    0.111307   
149 -1.644048       -1.310043             -0.537820   -0.860662   
150 -0.261328       -1.310043             -0.582029    0.186074   
151  0.618586        0.727802             -0.471507    0.335608   
152  0.555735       -1.310043             -0.360984    0.298224   

     serum_creatinine    sodium  potassium  hemoglobin  packed_cell_volume  \
2            1.683685 -0.956087   1.367853   -2.902248           -2.894560   
5            1.213526  0.447053  -1.298

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [ ]:
# Converted SQL to Pandas code

# count_df = dataset.groupby('Target').size().reset_index(name='count')
# print(count_df)

# Count the number of rows for each target
count_ckd = dataset['Target_chronic_kidney_disease'].sum()
count_not_ckd = dataset['Target_not_chronic_kidney_disease'].sum()
count_df = pd.DataFrame({
    'Target': ['not_ckd', 'ckd'],
    'count': [count_not_ckd, count_ckd]
})
print(count_df)


    Target  count
0  not_ckd    111
1      ckd     22


*Describe what the above code does here*

This code groups the dataset rows by the values in the target column. It creates 2 groups for "ckd" and "not_ckd". Then, it calculates the size (number of rows) with the target "ckd" and "not_ckd". Then, it resets the index and turns the output into a pandas dataframe. The dataframe will have 2 columns:

i. Target: "ckd" or "not_ckd".

ii. count: The number of elements with the target "ckd" and "not_ckd".

The single line of code is: count_df = dataset.groupby('Target').size().reset_index(name='count')

However, this does not work because we have 2 columns for "Target_ckd" and "Target_notckd". Thus, I revised the code using 383GPT to get the intended dataframe as the result.